machine/v1alpha1/provider_types.go

```diff
@@ -0,0 +1,207 @@
+	// The contents should be in the form of key:value. This secret must be in
+	// the same namespace as the provider.
+	// +optional
+	SecretName *string `json:"secretName,omitempty"`
+
+	// FetchConfig determines how the operator will fetch the components and metadata for the provider.
+	// If nil, the operator will try to fetch components according to default
+	// embedded fetch configuration for the given kind and `ObjectMeta.Name`.
+	// For example, the infrastructure name `aws` will fetch artifacts from
+	// https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases.
+	// +optional
+	FetchConfig *FetchConfiguration `json:"fetchConfig,omitempty"`
+}
+
+// ManagerSpec defines the properties that can be enabled on the controller manager for the provider.
+type ManagerSpec struct {
+	// ControllerManagerConfigurationSpec defines the desired state of GenericControllerManagerConfiguration.
+	ctrlconfigv1.ControllerManagerConfigurationSpec `json:",inline"`
+
+	// ProfilerAddress defines the bind address to expose the pprof profiler (e.g. localhost:6060).
+	// Default empty, meaning the profiler is disabled.
+	// Controller Manager flag is --profiler-address.
+	// +optional
+	ProfilerAddress *string `json:"profilerAddress,omitempty"`
+
+	// MaxConcurrentReconciles is the maximum number of concurrent Reconciles
+	// which can be run. Defaults to 10.
+	// +optional
+	// +kubebuilder:default=10
+	// +kubebuilder:validation:Minimum=1
+	MaxConcurrentReconciles *int `json:"maxConcurrentReconciles,omitempty"`
+
+	// Verbosity set the logs verbosity. Defaults to 1.
+	// Controller Manager flag is --verbosity.
+	// +optional
+	// +kubebuilder:default=1
+	// +kubebuilder:validation:Minimum=0
+	Verbosity int `json:"verbosity,omitempty"`
+
+	// Debug, if set, will override a set of fields with opinionated values for
+	// a debugging session. (Verbosity=5, ProfilerAddress=localhost:6060)
+	// +optional
+	// +kubebuilder:default=false
+	Debug bool `json:"debug,omitempty"`
+
+	// FeatureGates define provider specific feature flags that will be passed
+	// in as container args to the provider's controller manager.
+	// Controller Manager flag is --feature-gates.
+	FeatureGates map[string]bool `json:"featureGates,omitempty"`
+}
+
+// DeploymentSpec defines the properties that can be enabled on the Deployment for the provider.
+type DeploymentSpec struct {
+	// Number of desired pods. This is a pointer to distinguish between explicit zero and not specified. Defaults to 1.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	Replicas *int `json:"replicas,omitempty"`
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
```

Comment:
Is this a generic API or one for openshift?  I wouldn't expect users to be allowed to set this value on openshift.  Same for tolerations or affinity.

Reply:
This is a generic API, primarily designed to work with vanilla k8s but we want to use it as an "internal" API in openshift. It will not be documented to be used by customers and cluster admins will not find much use of it. There are 2 reasons for us to have it: 1. it makes integration of new ClusterAPI providers easier, the CRs will only be created and managed by openshift operators and it provides a generic for managing providers, additionally, if we ever need to run multiple providers on the management cluster there is no need to invent new logic. 2. It takes into account the new install flexibility effort, where this API allows a good 3rd party provider integration.